rust/src/server.rs

```diff
@@ -0,0 +1,190 @@
+use hyper::{client::HttpConnector, Body, Client, HeaderMap, Request, Response, StatusCode};
+use std::{convert::Infallible, net::SocketAddr};
+
+#[derive(Clone, Debug)]
+pub struct ProxyClient {
+    addr: SocketAddr,
+    forward_addr: SocketAddr,
+    http_client: Client<HttpConnector>,
+}
+
+impl ProxyClient {
+    pub fn new(addr: SocketAddr, forward_addr: SocketAddr) -> ProxyClient {
+        let http_client = Client::new();
+        return ProxyClient {
+            addr,
+            forward_addr,
+            http_client,
+        };
+    }
+    pub fn addr(&self) -> SocketAddr {
+        self.addr
+    }
+}
+
+#[macro_export]
+macro_rules! new {
+    ($e:expr) => {{
+        use crate::server::handle;
+        use hyper::{
+            service::{make_service_fn, service_fn},
+            Server,
+        };
+        type GenericError = Box<dyn std::error::Error + Send + Sync>;
+        let proxy_client: ProxyClient = $e;
+        let proxy_addr = proxy_client.addr();
+        let new_service = make_service_fn(move |_| {
+            let proxy_client = proxy_client.clone();
+            async {
+                Ok::<_, GenericError>(service_fn(move |req| {
+                    // Clone again to ensure that client outlives this closure.
+                    handle(req, proxy_client.to_owned())
+                }))
+            }
+        });
+        let builder = Server::bind(&proxy_addr);
+        builder.serve(new_service)
+    }};
+}
+
+pub(crate) use new;
+
+pub async fn handle(req: Request<Body>, proxy: ProxyClient) -> Result<Response<Body>, Infallible> {
+    let uri_string = format!("http://{}{}", proxy.forward_addr, req.uri().path());
+    let uri = uri_string
+        .parse::<hyper::Uri>()
+        .expect("proxy addr should parse");
+    let http_req = Request::builder()
+        .method(req.method())
+        .header(
+            "content-type",
+            get_header(req.headers(), "content-type", "text/plain"),
+        )
+        .header("DD-API-KEY", get_header(req.headers(), "DD-API-KEY", ""))
+        .header(
+            "user-agent",
+            get_header(req.headers(), "user-agent", "proxy-filter"),
+        )
+        .uri(uri)
+        .body(req.into_body());
+    if http_req.is_err() {
+        // let mut response = Response::new(Body::try_from(http_req.unwrap_err()).unwrap());
+        let mut response = Response::new(Body::empty());
+        *response.status_mut() = StatusCode::BAD_REQUEST;
+        return Ok(response);
+    }
+
+    let http_resp = proxy.http_client.request(http_req.unwrap()).await.unwrap();
```

Comment:
The
```rust
if http_req.is_err() {
    // ...
    return ..
}
... http_req.unwrap() .. 
```
could be better expressed  (without `unwrap`) with a match
```rust
match http_req {
    Error(_) => { ... }
    Ok(http_req) => { ... }
}
```
or alternately, if you want to keep the subsequent code at the same indentation level
```rust
let http_req = if let Ok(r) = http_req {
    r
} else {
    // ..error
}
```